File: Cpp/secante.cpp
// Se importan las librerias a utilizar
#include <iostream>
#include <math.h>
#include <vector>
#include "matplotlibcpp.h"

using namespace std;
namespace plt = matplotlibcpp;

// Definicion de la funcion del problema
double funcion(double x) {

    return exp(-pow(x,2)) - x; // e^(-(x^2)) - x = 0

}

double secante(double x0, double x1, double tol, int iterMax) {

    /*
        Esta funcion aproxima numericamente la solucion de una ecuacion no lineal
        por medio del metodo de la secante partiendo de dos valores iniciales hasta
        encontrar el cero de la funcion en cuestion.

        Parametros de entrada: x0 corresponde al primer valor inicial necesario para
                               comenzar el metodo iterativo.

                               x1 representa el segundo valor inicial necesario para
                               comenzar el metodo iterativo.

                               tol hace referencia al valor de tolerancio del error
                               absoluto.

                               iterMax es la cantidad maxima de iteraciones que realizara
                               el metodo para aproximar el cero de la funcion dada.

        Parametros de salida: aprox corresponde a la aproximacion del cero de la funcion.

                              error representa el error absoluto del cero aproximado.
                                
    */

    vector < double > iterVector; // Vector de iteraciones
    vector < double > errorVector; // Vector de errores

    double aprox;
    double error;
    double x_k_anterior = x0;
    double x_k_actual = x1;
    double x_k_siguiente;

    for (int k = 0; k < iterMax; k++) {

        x_k_siguiente = x_k_actual - ((x_k_actual - x_k_anterior)*funcion(x_k_actual)) /
                        (funcion(x_k_actual) - funcion(x_k_anterior));

        aprox = x_k_siguiente;

        error = abs(funcion(x_k_siguiente));

        iterVector.push_back(k);

        errorVector.push_back(error);

        if (error < tol) { // Condicion de parada del ciclo iterativo

            break;

        }

        x_k_actual = x_k_siguiente;

    }

    // Impresion de resultados de Iteraciones vs Error Absoluto del Metodo de la Secante
    cout << "Iteraciones = " << iterVector.back() << endl;
    cout << "Aproximacion = " << x_k_siguiente << endl;
    cout << "Error absoluto = " << error << endl;
    
    // Configuracion del grafico 
    plt::named_plot("Error Absoluto", iterVector, errorVector);
    plt::title("Grafico de Iteraciones vs Error Absoluto del Metodo de la Secante");
    plt::legend();
    plt::show();

    return x_k_siguiente;

}

int main() {

    // Ejemplo numerico
    double x0 = 0;
    double x1 = 1;
    double tol = pow(10,-3);
    int iterMax = 1000;

    double aprox = secante(x0, x1, tol, iterMax);

    return 0;
}

File: Octave/primer_examen/secante.m

function secante()
  
  % ejemplo del metodo de la secante
  
  clc; clear;
  
  f = 'exp(x)+x+2';
  x0 = 0;
  x1 = 1;
  tol = 10^-9;
  iterMax = 1000;
  
  [xk_1 error] = metodo_secante(f, x0, x1, tol, iterMax)
  
end

function [xk_1 error] = metodo_secante(f, x0, x1, tol, iterMax)
  
  pkg load symbolic;
  syms x;
  
  f1 = matlabFunction(sym(f));
  
  xk_0 = x0;
  xk_1 = x1;
  
  for k=0:iterMax
    
    n = (xk_1 - xk_0) * f1(xk_1);
    d = f1(xk_1) - f1(xk_0);
    
    xk_0 = xk_1;
    
    xk_1 = xk_1 - n/d;
    
    error = abs(f1(xk_1))
    
    if error < tol
      break
    endif
    
  endfor
  
end